clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1034,28 +1024,35 @@ private Record nextFetchedRecord() {
                     records = currentBatch.streamingIterator();
                 }
 
-                Record record = records.next();
-                maybeEnsureValid(record);
-
-                // skip any records out of range
-                if (record.offset() >= nextFetchOffset) {
-                    nextFetchOffset = record.offset() + 1;
-
-                    // control records are not returned to the user
-                    if (!currentBatch.isControlBatch())
-                         return record;
+                // If an exception was thrown from the last record. We should throw the same exception here again.
+                // We do not cache the exception thrown last time because the stack trace could be different
+                // from last time.
+                if (hasExceptionInLastFetch) {
+                    maybeEnsureValid(lastRecord);
```

Comment:
Yes, you are right. I was thinking that we may need to give it a couple of retries and a back off for each retry then throw exception. But it seems that we can just retry immediately and only once. If that is the case, we don't have to worry about the backoff/retries.